// vuetify/lib/composables/loader.mjs
import { computed, createVNode } from "vue";
import { VProgressLinear } from "../components/VProgressLinear/VProgressLinear.mjs";
import { propsFactory } from "../util/propsFactory.mjs";
import { getCurrentInstanceName } from "../util/getCurrentInstance.mjs";
const makeLoaderProps = propsFactory({
  loading: [Boolean, String]
}, "loader");
function useLoader(props) {
  let name = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : getCurrentInstanceName();
  const loaderClasses = computed(() => ({
    [`${name}--loading`]: props.loading
  }));
  return {
    loaderClasses
  };
}
function LoaderSlot(props, _ref) {
  var _a;
  let {
    slots
  } = _ref;
  return createVNode("div", {
    "class": `${props.name}__loader`
  }, [((_a = slots.default) == null ? void 0 : _a.call(slots, {
    color: props.color,
    isActive: props.active
  })) || createVNode(VProgressLinear, {
    "absolute": props.absolute,
    "active": props.active,
    "color": props.color,
    "height": "2",
    "indeterminate": true
  }, null)]);
}
export {
  LoaderSlot,
  makeLoaderProps,
  useLoader
};
